driver-core/src/main/java/com/datastax/driver/core/NettyUtil.java

```diff
@@ -0,0 +1,156 @@
+import io.netty.channel.nio.NioEventLoopGroup;
+import io.netty.channel.socket.SocketChannel;
+import io.netty.channel.socket.nio.NioSocketChannel;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * A set of utilities related to the underlying Netty layer.
+ */
+@SuppressWarnings("unchecked")
+class NettyUtil {
+
+    private static final boolean FORCE_NIO = SystemProperties.getBoolean("com.datastax.driver.FORCE_NIO", false);
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(NettyUtil.class);
+
+    private static final boolean SHADED;
+
+    private static final boolean USE_EPOLL;
+
+    private static final Constructor<? extends EventLoopGroup> EPOLL_EVENT_LOOP_GROUP_CONSTRUCTOR;
+
+    private static final Class<? extends SocketChannel> EPOLL_CHANNEL_CLASS;
+
+    private static final Class[] EVENT_GROUP_ARGUMENTS = { int.class, ThreadFactory.class };
+
+    static {
+        boolean shaded;
+        try {
+            // prevent this string from being shaded
+            Class.forName(String.format("%s.%s.channel.Channel", "io", "netty"));
+            shaded = false;
+        } catch (ClassNotFoundException e) {
+            try {
+                Class.forName("com.datastax.shaded.netty.channel.Channel");
+                shaded = true;
+            } catch (ClassNotFoundException e1) {
+                throw new AssertionError("Cannot locate Netty classes in the classpath:" + e1);
+            }
+        }
+        SHADED = shaded;
+        boolean useEpoll = false;
+        try {
+            Class<?> epool = Class.forName("io.netty.channel.epoll.Epoll");
+            if (FORCE_NIO) {
+                LOGGER.info("Found Netty's native epoll transport in the classpath, "
+                    + "but NIO was forced through the FORCE_NIO system property.");
+            } else if (!System.getProperty("os.name", "").toLowerCase(Locale.US).equals("linux")) {
+                LOGGER.warn("Found Netty's native epoll transport, but not running on linux-based operating " +
+                    "system. Using NIO instead.");
+            } else if (!(Boolean)epool.getMethod("isAvailable").invoke(null)) {
+                LOGGER.warn("Found Netty's native epoll transport in the classpath, but epoll is not available. "
+                    + "Using NIO instead.", (Throwable)epool.getMethod("unavailabilityCause").invoke(null));
+            } else {
+                LOGGER.info("Found Netty's native epoll transport in the classpath, using it");
+                useEpoll = true;
+            }
+        } catch (ClassNotFoundException e) {
+            LOGGER.info("Did not find Netty's native epoll transport in the classpath, defaulting to NIO.");
+        } catch (Exception e) {
+            LOGGER.warn("Unexpected error trying to find Netty's native epoll transport in the classpath, defaulting to NIO.", e);
+        }
+        USE_EPOLL = useEpoll;
+        Constructor<? extends EventLoopGroup> constructor = null;
```

Comment:
And this is the code to locate epoll classes without importing them (previously in NettyOptions).